### ISSUE TITLE
Is it possible to enable gradients when I make inference with an image in a trained model? I am trying to compute the saliency map.

### ISSUE THREAD

**Issue by AntonioVispi** [open-mmlab/mmclassification#1151]
### Branch

master branch (0.24 or other 0.x version)

### Describe the feature

Greetings,
I am trying to implement the saliency map with a model that I have properly trained. The problem I am experiencing is due to an incompatibility which I cannot manage. To implement the saliency map I need the inference image to have the `requires_grad = True` flag enabled, however, if I make inference with the `inference_model(model,image)` function, I can input only ndarray images, which do not have the `require_grad` flag, to as far as I know. If I load the image as follows 
`from PIL import Image`
`img = Image.open ('/PATH')`
 I am able to enable the `requires_grad` flag, however I cannot inference with the `inference_model` function, as it takes ndarray and not Tensors as input.
In summary, I can make the inference correctly, I can compute `backward()`, but since I have not set the `requires_grad = True` flag, I cannot see the gradient generated by my image, and therefore I cannot create the saliency map.

I emphasize that if I could solve this problem it would be a pleasure to eventually help add a Tool to your repository.
The site from which I took inspiration for the implementation of the saliency map is the following: [https://github.com/sunnynevarekar/pytorch-saliency-maps.git](url)

Thanks in advance, I hope you will help me.

### Will you implement it?

- [X] I would like to implement this feature and create a PR!

**Comment by Ezra-Yu** [comment]
Here is the code, I modify from the [demo.py](https://github.com/open-mmlab/mmclassification/blob/master/demo/image_demo.py)

```
# Copyright (c) OpenMMLab. All rights reserved.
from argparse import ArgumentParser

import mmcv
import torch

from mmcls.apis import inference_model, init_model, show_result_pyplot


def main():
    parser = ArgumentParser()
    # parser.add_argument('img', help='Image file')
    parser.add_argument('config', help='Config file')
    parser.add_argument('checkpoint', help='Checkpoint file')
    parser.add_argument(
        '--show',
        action='store_true',
        help='Whether to show the predict results by matplotlib.')
    parser.add_argument(
        '--device', default='cuda:0', help='Device used for inference')
    args = parser.parse_args()

    # build the model from a config file and a checkpoint file
    model = init_model(args.config, args.checkpoint, device=args.device)
    x = torch.rand( (1, 3, 224, 224) )
    x.requires_grad = True
    y = model(x, return_loss=False, softmax=False, post_process=False)
    print(type(y), y.size())
    score, indices = torch.max(y, 1)
    score.backward()
    slc, _ = torch.max(torch.abs(x.grad[0]), dim=0)
    #normalize to [0..1]
    slc = (slc - slc.min())/(slc.max()-slc.min())
    print(slc.size())


    # test a single image
    #result = inference_model(model, args.img)
    # show the results
    # print(mmcv.dump(result, file_format='json', indent=4))
    # if args.show:
    #    show_result_pyplot(model, args.img, result)


if __name__ == '__main__':
    main()
```

use 
```python demo/image_demo.py configs/resnet/resnet50_8xb32_in1k.py https://download.openmmlab.com/mmclassification/v0/resnet/resnet50_8xb32_in1k_20210831-ea4938fc.pth  --device cpu```

it works fine.

**Comment by Ezra-Yu** [comment]
Is this method the same with the [`torch.Captum`](https://captum.ai/tutorials/Resnet_TorchVision_Interpret)?
There is a [issue](https://github.com/open-mmlab/mmclassification/issues/910) to integrat some tools of Captum. May it possible to just use that from the Captum?

**Comment by AntonioVispi** [comment]
It worked great! I actually made some changes. So I will summarize the steps I have taken if someone needs another way to do it. Thanks really anyway. I have not tried with Captum
`model = init_model(cfg, checkpoint_file, device) # Inizialize the model with the same training cfg`
`immagine = ('/PATH') # Full path of the image we want the saliency map of`
`device = 'cpu'`
### inference_model changed to output the image exactly as the model expects it to be

```
from mmcv.parallel import collate, scatter
from mmcls.datasets.pipelines import Compose
import numpy as np

def inference_model_mine(model, img):
    """Inference image(s) with the classifier.
    Args:
        model (nn.Module): The loaded classifier.
        img (str/ndarray): The image filename or loaded image.
    Returns:
        result (dict): The classification results that contains
            `class_name`, `pred_label` and `pred_score`.
    """
    cfg = model.cfg
    device = next(model.parameters()).device  # model device
    # build the data pipeline
    if isinstance(img, str):
        if cfg.data.test.pipeline[0]['type'] != 'LoadImageFromFile':
            cfg.data.test.pipeline.insert(0, dict(type='LoadImageFromFile'))
        data = dict(img_info=dict(filename=img), img_prefix=None)
    else:
        if cfg.data.test.pipeline[0]['type'] == 'LoadImageFromFile':
            cfg.data.test.pipeline.pop(0)
        data = dict(img=img)
    test_pipeline = Compose(cfg.data.test.pipeline)
    data = test_pipeline(data)
    data = collate([data], samples_per_gpu=1)
    if next(model.parameters()).is_cuda:
        # scatter to specified GPU
        data = scatter(data, [device])[0]

    return data['img']
```
#####

So run the modified function:

`img = inference_model_mine(model, immagine)`

```
from PIL import Image
from argparse import ArgumentParser
import mmcv
import torch
from mmcls.apis import inference_model, init_model, show_result_pyplot

img.requires_grad = True
y = model(img, return_loss=False,  post_process=False)
print(type(y), y.size())
score, indices = torch.max(y, 1)
score.backward()
slc, _ = torch.max(torch.abs(img.grad[0]), dim=0)
#normalize to [0..1]
slc = (slc - slc.min())/(slc.max()-slc.min())
print(slc.size())
```
### Your Saliency map is done!
